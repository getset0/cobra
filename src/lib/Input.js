import styled, { css } from "styled-components";
import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from "@fortawesome/free-solid-svg-icons";

// the width of the input border is controled by the padding attribute
const InputWrapperSpan = styled.span`
  position:relative;
  margin: 7px 4px;
  background: linear-gradient(21deg, ${props => props.theme.darkestBlueColor}, ${props => props.theme.lightBlueColor});
  padding: 3px;
  display: inline-block;
  border-radius: 9999em;

  ${props => props.disabled && css`
      background: ${props => props.theme.lightGrayColor};
  `}

`;


// margin should be a bit bigger then InputWrapperSpan padding
const InputInnerSpan = styled.span`
    transform: scale(.993, .94);
    transition: transform .3s, opacity .2s;
    opacity: 0;

    position:absolute;
    z-index: 0;
    margin: 4px;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
    border-radius: inherit;
    pointer-events: none;

    box-shadow: inset 0 0 0 3px #fff,
      0 0 0 4px #fff,
      3px -3px 22px ${props => props.theme.darkestBlueColor},
      -3px 3px 22px ${props => props.theme.lightBlueColor};

    ${props => props.disabled && css`
      box-shadow: 0;
    `}

`;

// z-index should be above InputInnerSpan
const Input = styled.input`
  position: relative;
  display: inherit;
  border-radius: inherit;
  line-height: 1.5em;
  margin: 0;
  width: ${props => props.width};
  height: ${props => props.height};
  border: none;
  outline: none;
  font: 300 1em system-ui;
  padding: .3em .75em;
  z-index: 1;
  ${props => props.type === "search" && css`
    padding: .3em .75em .3em 1.7em;
  `}

  &:focus + span {
    opacity: 1;
    transform: scale(1);
  }
`;

const InputIconWrapper = styled.span`
  position: absolute;
  width: 14px;
  left: 9px;
  top: 9px;
  right: 0;
  bottom: 0;
  z-index: 2;
`;

const CustomInput = ({ disabled, ...props }) => (
  <InputWrapperSpan disabled={disabled}>
    {props.type === "search" &&
      <InputIconWrapper>
        <FontAwesomeIcon icon={faSearch} />
      </InputIconWrapper>
    }
    <Input disabled={disabled} {...props} />
    <InputInnerSpan disabled={disabled} />
  </InputWrapperSpan>
);

export default CustomInput;
