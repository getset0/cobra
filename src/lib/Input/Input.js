import styled, { css } from "styled-components";
import { theme } from "../constants";
import React from "react";
import PropTypes from "prop-types";

const InputWrapperSpan = styled.span`
    display: flex;
    flex-direction: column;
    width: ${props => props.width}px;
`;
const Input = styled.input`
    padding: 2px 14px 4px;
    font-size: 18px;
    ${props =>
    props.error ? css`
            border: 1px solid lightcoral;
    ` : props.valid ? css`
            border: 1px solid ${theme.greenColor}
    ` : ""}
`;
const LabelSpan = styled.span`
    padding: 4px 0;
    font-size: 14px;
    font-weight: bold;
    color: ${theme.mediumGrayColor}
`;

const Errorspan = styled.span`
    padding: 2px 0;
    color: lightcoral;
    font-size: 12px;
`;

const CustomInput = ({
  label,
  valid,
  error,
  width,
  ...props
}) => (
  <InputWrapperSpan width={width}>
    {label && <LabelSpan>{label}</LabelSpan>}
    <Input valid={valid} error={error} {...props} />
    {error && <Errorspan>{error}</Errorspan>}
  </InputWrapperSpan>
);

CustomInput.propTypes = {
  label: PropTypes.oneOf([ PropTypes.string, PropTypes.element ]),
  error: PropTypes.oneOf([ PropTypes.string, PropTypes.element ]),
  valid: PropTypes.bool,
  width: PropTypes.number
};

CustomInput.defaultProps = {
  width: 240
};

export default CustomInput;
