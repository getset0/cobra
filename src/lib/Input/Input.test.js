import React from "react";
import Input from "./Input";
import "jest-styled-components";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

it("renders Input without crashing", () => {
  shallow(<Input />);
});

it("renders normal Input correctly", () => {
  const wrapper = shallow(<Input />);
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("correctly renders disabled input", () => {
  const wrapper = shallow(
    <Input placeholder="test" disabled />
  );
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("correctly renders search input", () => {
  const wrapper = shallow(
    <Input placeholder="test" type="search" />
  );
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("correctly renders password input", () => {
  const wrapper = shallow(
    <Input placeholder="test" type="password" />
  );
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("correctly renders password input", () => {
  const wrapper = shallow(
    <Input placeholder="test" type="number" />
  );
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});
