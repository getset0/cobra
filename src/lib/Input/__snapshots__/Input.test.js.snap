// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders disabled input 1`] = `
<styled.span
  disabled={true}
>
  <styled.input
    disabled={true}
    placeholder="test"
    theme={
      Object {
        "blackColor": "#000000",
        "darkBlueColor": "#02469B",
        "darkGrayColor": "#4A4A4A",
        "darkestBlueColor": "#02254D",
        "darkestGrayColor": "#303030",
        "greenColor": "#76B041",
        "lightBlueColor": "#0190D0",
        "lightGrayColor": "#D8D8D8",
        "lightestBlueColor": "#00FEFE",
        "mediumGrayColor": "#979797",
        "primaryColor": "#1676B7",
        "secondaryColor": "#17BEBB",
        "whiteBackgroundColor": "#fdfdfd",
        "whiteColor": "#ffffff",
        "yellowColor": "#FFC914",
      }
    }
  />
  <styled.span
    disabled={true}
  />
</styled.span>
`;

exports[`correctly renders password input 1`] = `
<styled.span
  disabled={false}
>
  <styled.input
    disabled={false}
    placeholder="test"
    theme={
      Object {
        "blackColor": "#000000",
        "darkBlueColor": "#02469B",
        "darkGrayColor": "#4A4A4A",
        "darkestBlueColor": "#02254D",
        "darkestGrayColor": "#303030",
        "greenColor": "#76B041",
        "lightBlueColor": "#0190D0",
        "lightGrayColor": "#D8D8D8",
        "lightestBlueColor": "#00FEFE",
        "mediumGrayColor": "#979797",
        "primaryColor": "#1676B7",
        "secondaryColor": "#17BEBB",
        "whiteBackgroundColor": "#fdfdfd",
        "whiteColor": "#ffffff",
        "yellowColor": "#FFC914",
      }
    }
    type="password"
  />
  <styled.span
    disabled={false}
  />
</styled.span>
`;

exports[`correctly renders password input 2`] = `
<styled.span
  disabled={false}
>
  <styled.input
    disabled={false}
    placeholder="test"
    theme={
      Object {
        "blackColor": "#000000",
        "darkBlueColor": "#02469B",
        "darkGrayColor": "#4A4A4A",
        "darkestBlueColor": "#02254D",
        "darkestGrayColor": "#303030",
        "greenColor": "#76B041",
        "lightBlueColor": "#0190D0",
        "lightGrayColor": "#D8D8D8",
        "lightestBlueColor": "#00FEFE",
        "mediumGrayColor": "#979797",
        "primaryColor": "#1676B7",
        "secondaryColor": "#17BEBB",
        "whiteBackgroundColor": "#fdfdfd",
        "whiteColor": "#ffffff",
        "yellowColor": "#FFC914",
      }
    }
    type="number"
  />
  <styled.span
    disabled={false}
  />
</styled.span>
`;

exports[`correctly renders search input 1`] = `
<styled.span
  disabled={false}
>
  <styled.span>
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f002",
            "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
          ],
          "iconName": "search",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </styled.span>
  <styled.input
    disabled={false}
    placeholder="test"
    theme={
      Object {
        "blackColor": "#000000",
        "darkBlueColor": "#02469B",
        "darkGrayColor": "#4A4A4A",
        "darkestBlueColor": "#02254D",
        "darkestGrayColor": "#303030",
        "greenColor": "#76B041",
        "lightBlueColor": "#0190D0",
        "lightGrayColor": "#D8D8D8",
        "lightestBlueColor": "#00FEFE",
        "mediumGrayColor": "#979797",
        "primaryColor": "#1676B7",
        "secondaryColor": "#17BEBB",
        "whiteBackgroundColor": "#fdfdfd",
        "whiteColor": "#ffffff",
        "yellowColor": "#FFC914",
      }
    }
    type="search"
  />
  <styled.span
    disabled={false}
  />
</styled.span>
`;

exports[`renders normal Input correctly 1`] = `
<styled.span
  disabled={false}
>
  <styled.input
    disabled={false}
    theme={
      Object {
        "blackColor": "#000000",
        "darkBlueColor": "#02469B",
        "darkGrayColor": "#4A4A4A",
        "darkestBlueColor": "#02254D",
        "darkestGrayColor": "#303030",
        "greenColor": "#76B041",
        "lightBlueColor": "#0190D0",
        "lightGrayColor": "#D8D8D8",
        "lightestBlueColor": "#00FEFE",
        "mediumGrayColor": "#979797",
        "primaryColor": "#1676B7",
        "secondaryColor": "#17BEBB",
        "whiteBackgroundColor": "#fdfdfd",
        "whiteColor": "#ffffff",
        "yellowColor": "#FFC914",
      }
    }
  />
  <styled.span
    disabled={false}
  />
</styled.span>
`;
